README.md
BearLibTerminal provides a pseudoterminal window with a grid of character cells and a simple yet powerful API for flexible textual output and uncomplicated input processing.

### Features

* Unicode support: you can use UTF-8 or UTF-16/32 wide strings easily.
* Support for bitmap and vector (TrueType) fonts and tilesets.
* Extended output facilities: tile composition, alignment, offsets.
* High performance (uses OpenGL).
* Keyboard and mouse support.
* Bindings for several programming languages: С/С++, C#, Go, Lua, Pascal, Python, Ruby.
* Windows, Linux and OS X support.


### Documentation

* [Reference](http://foo.wyrd.name/en%3Abearlibterminal%3Areference) (functions, constants, behaviour)
* [Design overview](http://foo.wyrd.name/en%3Abearlibterminal%3Adesign)


### Using

Some notes about using it with various languages or compilers:


#### C/C++

Visual C++ projects should be linked against BearLibTerminal.lib import library (specify it in the additional linker dependencies).

MinGW projects should link against .dll directly (the .lib is just an import library for Visual C++, do not copy it):

    g++ -I/path/to/header -L/path/to/dll main.cpp -lBearLibTerminal -o app.exe


#### Python

Python packages are available. You can install everything you need to use the library with

    pip install bearlibterminal

Depending on the OS and Python installation, you might also want to

- Replace `pip` with `pip3` or `python3 -m pip` to select a correct version of Python.
- Add `--user` flag to install package locally (i. e. user-wide).

Package contains both wrapper and an appropriate binary for the platform, so you do not need to copy anything else anywhere. Just import the library in the source:

    from bearlibterminal import terminal
    
    terminal.open()
    terminal.printf(2, 1, "Hello, world!")
    terminal.refresh()
    while terminal.read() != terminal.TK_CLOSE:
        pass
    terminal.close()

#### Lua

Wrapper for Lua is built-in. You need to use a regular (dynamic) Lua runtime and place BearLibTerminal binary
in a suitable location (e. g. in the same directory as script). For Linux you'll also need to rename the .so
to just 'BearLibTerminal.so' (dropping the 'lib' prefix). After that it would be possible to import the library
the usual way:

    local terminal = require "BearLibTerminal" 

### Building

BearLibTerminal is a language-agnostic dynamic-link library (.dll/.so/.dylib), therefore you generally do not have
to build it yourself and may simply use the [prebuilt binaries](http://foo.wyrd.name/en:bearlibterminal#download).

To build BearLibTerminal you will need CMake and a recent GCC/MinGW compiler. For Linux any GCC version 4.6.3 and above
will do. For Windows there are several MinGW builds with various quirks, using [TDM-GCC](http://tdm-gcc.tdragon.net/) or
[mingw-builds](http://mingw-w64.org/doku.php/download/mingw-builds) (a flavour of mingw-w64) is recommended.
MinGW compiler MUST use Posix thread model.


### License

The library is licensed mainly under the MIT license with a few parts under other permissive licenses.

CHANGELOG.md
### 0.15.8 (2022-12-06)

- Add support for drawing NumPy structured arrays.

### 0.15.7 (2017-11-20)

- Relax restriction on maximum terminal size from 256x256 to 1024x1024 cells. 
- Add `output.texture-filter` option (value is `linear` or `nearest`).
- Make output.texture-filter toggleable at runtime, e. g. by Alt+F.

### 0.15.5 (2017-11-13)

- Fix background color for tiles with spacing bigger than 1x1.
- Fix char type signedness (resolves [issue #47](https://bitbucket.org/cfyzium/bearlibterminal/issues/47/g-arm-default-char-is-not-signed)).
- Add missing include to sample's Main.cpp (resolves [issue #48](https://bitbucket.org/cfyzium/bearlibterminal/issues/48/sample-program-doesnt-build-on-osx)).

### 0.15.4 (2017-06-30)

- Fix Alt key being stuck after regaining focus (e. g. after Alt-Tab).
- Fix blank line height/spacing in print (resolves [issue #41](https://bitbucket.org/cfyzium/bearlibterminal/issues/41/terminal_print-ignores-font-spacing-for)). 
- Fix memory leak in Cocoa event processing (resolves [issue #44](https://bitbucket.org/cfyzium/bearlibterminal/issues/44/possible-memory-issue-with-terminalread)).
- Fix crashing when being opened after SDL (e. g. after libtcod import in Python, see [issue #42](https://bitbucket.org/cfyzium/bearlibterminal/issues/42/improved-error-handling-with-sdl-errors)).

### 0.15.3 (2017-05-14)

- Fix adding sprite tiles to an atlas (closes issue #30).
- Fix excess generation of 'character replacement' tiles.
- Fix dynamic tile generation (individual to a font, closes issue #32).
- Fix configuring several fonts in one `set()` call (closes issue #34).
- Add `font()` function (select the current font by name, closes issue #33).

### 0.15.2 (2017-03-12)

- Retrieve clipboard contents via `terminal_get("clipboard")`.
- Fix bitmap tileset reverse codepage (sparse tileset) handling.
- `input.cursor-blink-rate=0` disables cursor blinking in `terminal_read_str()`. 
- Fix printing tab characters (configured by `output.tab-width` option).
- Fix app hanging on reopening terminal in macOS (see [issue #23](https://bitbucket.org/cfyzium/bearlibterminal/issues/23/reopening-a-closed-terminal-hangs-forever)).

### 0.15.1 (2017-01-18)

- Change default hinting to font's native hinter.
- Add `hinting` TrueType font attribute.
- Make Python wrapper look for the library binary near the executable (should help packaging the application). 
- Minor fix in the sample application.

### 0.15.0 (2017-01-09)

- Add bbox and alignment parameters to `print()` function (instead of using in-string formatting tags).
- Add bbox parameters to `measure()` function.
- Add `dead-center` tile alignment (centering by image's center of mass).
- Add `use-box-drawing` and `use-block-elements` truetype font attributes (disables auto-generation of respective characters).
- Add text resource descriptors (specifying resources inside a configuration string, e. g. codepages).
- Add version property (available via `get()`).
- Fix TrueType font alignment at arbitrary tile size.
- Fix minor memory leaks (in X11Window and TrueType font constructors).
- Fix keybad keys not producing character codes (`TK_CHAR`/`TK_WCHAR` states).

### 0.14.12 (2016-12-01)

- Fix mapping of first 32 characters of CP437 codepage.
- Fix priority of auto-generated Box Drawing tiles.
- Fix crashing Python interpeter when exiting suddenly.
- Add support for grayscale bitmap font images w/o transparency.
- Setting a font without specifying a size parameter is not allowed anymore.
- C#: minor wrapper improvements (function overloads and Size type support in Set/Get).

### 0.14.11 (2016-10-30)

- Fix 'resize' parameter in bitmap tilesets (also change its meaning to target tile size).
- Fix toggling fullscreen in Linux (_DIALOG window type is incompatible with _FULLSCREEN).
- Fix crash under Wine (dynamic library was being unloaded way too early).
- Fix crash on the older hardware (sprite textures were not honoring the lack of NPOTD support).
- Fix out-of-bounds reads in bilinear filter (produced artifacts on bitmap borders).
- Add missing TK\_MOUSE\_SCROLL constant in Lua binding.

### 0.14.10 (2016-10-16)

- The window is now centered on a screen at startup (see [issue #15](https://bitbucket.org/cfyzium/bearlibterminal/issues/15/os-x-window-isnt-centered-in-the-screen)).
- Add Cmd+Q app menu for OS X ([issue #16](https://bitbucket.org/cfyzium/bearlibterminal/issues/16/os-x-command-q-doesnt-work)).
- Fix first redraw on OS X 10.11 ([issue #14](https://bitbucket.org/cfyzium/bearlibterminal/issues/14/loading-takes-a-long-time-and-cant-be)).
- Fix flooding mouse-movement events (again, finally).
- Python wrapper fixes (performance optimizations, check() behaviour, a bug in get(); see [issue #17](https://bitbucket.org/cfyzium/bearlibterminal/issues/17/python-optimizations), [issue #8](https://bitbucket.org/cfyzium/bearlibterminal/issues/8/check-bug-in-the-python-wrapper)).
- Fix dynamically generated character U+2523.

### 0.14.8 (2016-09-06)

- Setting custom colors in the palette (available later via `color_from_name()` and the `[color=name]` formatting tag)
- Specifying custom colors in the 'Palette' section of a configuration file.
- Fix filtered events ordering (some events could have been lost before).
- Pascal: Delphi-compatible Pascal wrapper.

### 0.14.7 (2016-08-31)

- Fixed parsing resource names (some absolute paths were incorrectly treated as memory addresses).

### 0.14.6 (2016-08-30)

- 'System' events like TK_CLOSE are always read regardless of input filter (no need to explicitly specify them while setting the filter).

### 0.14.5 (2016-08-29)

- Add support for Lua 5.3
- Fix reading some configuration options.
- Fix blocking on input when the window is not yet shown (would deadlock).
- Default input filter is 'keyboard, system' now (terminal read is equivalent to 'press any key').

### 0.14.4 (2016-08-17)

- Fix reading/updating single parameters in configuration file.

### 0.14.3 (2016-07-22)


Terminal/Source/Base64.cpp
/*
* BearLibTerminal
* Copyright (C) 2013 Cfyz
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
* of the Software, and to permit persons to whom the Software is furnished to do
* so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
* FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
* COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
* IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
* CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

#include <stdexcept>
#include "Base64.hpp"

namespace BearLibTerminal
{
	static const signed char b64_indexes[] = // TODO: reformat
	{
		/* 0 - 31 / 0x00 - 0x1f */
			-1, -1, -1, -1, -1, -1, -1, -1
		,   -1, -1, -1, -1, -1, -1, -1, -1
		,   -1, -1, -1, -1, -1, -1, -1, -1
		,   -1, -1, -1, -1, -1, -1, -1, -1
		/* 32 - 63 / 0x20 - 0x3f */
		,   -1, -1, -1, -1, -1, -1, -1, -1
		,   -1, -1, -1, 62, -1, -1, -1, 63  /* ... , '+', ... '/' */
		,   52, 53, 54, 55, 56, 57, 58, 59  /* '0' - '7'          */
		,   60, 61, -1, -1, -1,  0, -1, -1  /* '8', '9', ...      */
		/* 64 - 95 / 0x40 - 0x5f */
		,   -1, 0,  1,  2,  3,  4,  5,  6   /* ..., 'A' - 'G'     */
		,   7,  8,  9,  10, 11, 12, 13, 14  /* 'H' - 'O'          */
		,   15, 16, 17, 18, 19, 20, 21, 22  /* 'P' - 'W'          */
		,   23, 24, 25, -1, -1, -1, -1, -1  /* 'X', 'Y', 'Z', ... */
		/* 96 - 127 / 0x60 - 0x7f */
		,   -1, 26, 27, 28, 29, 30, 31, 32  /* ..., 'a' - 'g'     */
		,   33, 34, 35, 36, 37, 38, 39, 40  /* 'h' - 'o'          */
		,   41, 42, 43, 44, 45, 46, 47, 48  /* 'p' - 'w'          */
		,   49, 50, 51, -1, -1, -1, -1, -1  /* 'x', 'y', 'z', ... */

		,   -1, -1, -1, -1, -1, -1, -1, -1
		,   -1, -1, -1, -1, -1, -1, -1, -1
		,   -1, -1, -1, -1, -1, -1, -1, -1
		,   -1, -1, -1, -1, -1, -1, -1, -1

		,   -1, -1, -1, -1, -1, -1, -1, -1
		,   -1, -1, -1, -1, -1, -1, -1, -1
		,   -1, -1, -1, -1, -1, -1, -1, -1
		,   -1, -1, -1, -1, -1, -1, -1, -1

		,   -1, -1, -1, -1, -1, -1, -1, -1
		,   -1, -1, -1, -1, -1, -1, -1, -1
		,   -1, -1, -1, -1, -1, -1, -1, -1
		,   -1, -1, -1, -1, -1, -1, -1, -1

		,   -1, -1, -1, -1, -1, -1, -1, -1
		,   -1, -1, -1, -1, -1, -1, -1, -1
		,   -1, -1, -1, -1, -1, -1, -1, -1
		,   -1, -1, -1, -1, -1, -1, -1, -1
	};

	std::vector<std::uint8_t> Base64::Decode(const std::string& s)
	{
		if ( s.empty() )
		{
			return std::vector<std::uint8_t>();
		}

		if ( s.length()%4 != 0 )
		{
			throw std::invalid_argument("[Base64::Decode] Malformed string, not a multilply of four in length");
		}

		size_t size = (s.length()/4)*3;
		if ( s[s.length()-1] == '=' ) size -= 1;
		if ( s[s.length()-2] == '=' ) size -= 1;

		std::vector<std::uint8_t> result(size);

		auto b64code = [&](unsigned char c) -> unsigned char
		{
			signed char b;

			if ( c >= 0x7F || (b = b64_indexes[c]) == -1 )
			{
				throw std::invalid_argument("[Base64::Decode] Invalid character in string");
			}

			return (unsigned char)b;
		};

		size_t index = 0;

		for ( size_t i=0; i<s.length()/4; i++ )
		{
			std::uint32_t block = 0;
			
			for ( size_t j=0; j<4; j++ ) block = (block << 6) | (b64code(s[i*4+j]) & 0xFF);

			for ( size_t j=0; j<3 && index<size; j++ )
			{
				result[index++] = ((unsigned char*)&block)[2-j];
			}
		}

		return result;
	}
}
